src/eos-companion-app-service-managed-cache.c

```diff
@@ -0,0 +1,201 @@
+
+/**
+ * threadsafe_registered_cache_put:
+ * @cache: A #ThreadsafeRegisteredCache.
+ * @key: The key to use. If an entry was already in the cache with this key
+ *       it will be replaced.
+ * @value: (transfer full): The corresponding value.
+ *
+ * Insert a value into the cache with the lock taken.
+ */
+void
+threadsafe_registered_cache_put (ThreadsafeRegisteredCache *cache,
+                                 const gchar               *key,
+                                 gpointer                   value)
+{
+  g_mutex_lock (&cache->mutex);
+
+  g_hash_table_insert (cache->ht,
+                       g_strdup (key),
+                       value);
+
+  g_mutex_unlock (&cache->mutex);
+}
+
+/**
+ * threadsafe_registered_cache_lookup:
+ * @cache: A #ThreadsafeRegisteredCache.
+ * @key: The key to use.
+ * @value: (out): (transfer none): An outparam with the value which will be set
+ *                                 if it was in the cache. %NULL may be
+ *                                 returned if the underlying value was %NULL -
+ *                                 check the return value to see if the lookup
+ *                                 succeeded.
+ *
+ * Retrieve a value into from the cache with the lock taken.
+ *
+ * Returns: %TRUE if the value was in the cache, %FALSE otherwise.
+ */
+gboolean
+threadsafe_registered_cache_lookup (ThreadsafeRegisteredCache *cache,
+                                    const gchar               *key,
+                                    gpointer                  *value)
+{
+  gboolean ret = FALSE;
+
+  g_return_val_if_fail (value != NULL, FALSE);
+  g_mutex_lock (&cache->mutex);
+
+  ret = g_hash_table_lookup_extended (cache->ht,
+                                      key,
+                                      NULL,
+                                      value);
+
+  g_mutex_unlock (&cache->mutex);
+  return ret;
+}
+
+/**
+ * eos_companion_app_service_managed_cache_clear:
+ * @cache: A #EosCompanionAppServiceManagedCache.
+ *
+ * Clear all the entries in this #EosCompanionAppServiceManagedCache, forcing
+ * them to be regenerated if required.
+ */
+void
+eos_companion_app_service_managed_cache_clear (EosCompanionAppServiceManagedCache *cache)
+{
+  EosCompanionAppServiceManagedCachePrivate *priv = eos_companion_app_service_managed_cache_get_instance_private (cache);
+
+  threadsafe_registered_cache_clear (priv->registered_cache);
+}
+
+/**
+ * eos_companion_app_service_managed_cache_get_subcache: (skip)
+ * @cache: A #EosCompanionAppServiceManagedCache.
+ * @key: The key to lookup the subcache with.
+ * @value_destroy: A #GDestroyNotify for the value type of the subcache, if
+ *                 it needs to be created.
+ *
+ * Get a #ThreadsafeRegisteredCache from a #EosCompanionAppServiceManagedCache.
+ *
+ * Returns: (transfer none): A #ThreadsafeRegisteredCache.
+ */
+ThreadsafeRegisteredCache *
+eos_companion_app_service_managed_cache_get_subcache (EosCompanionAppServiceManagedCache *cache,
```

Comment:
Hmm, I'm not too keen on using singletons if we can avoid it, since singletons are really just more global state that need to be managed (eg, in the tests we have to set it up and tear it down). The dependency injection approach makes for annoying repetition, but it is cleaner overall...

> In other words, I'd expect either different parts of the to code create their own cache (what you have right now as keys to subcaches) and manage their lifecycle, or that there's a global cache in the process that different clients can always access and add their "namespace" in. Right now it's a mix of both, which is odd.

The latter is basically the way that it works right now. If a function wants to use some sort of global cache, then it takes an `EosCompanionAppIntegrationManagedCache *` object and registers its cache namespace in there. In general, functions shouldn't be creating locally managed static caches.

I think the confusion comes from the `ThreadsafeRegisteredCache` thing and its possible for us to just remove that and take a `GHashTable` directly. I'll check the thread-safety of `GHashTable`